specs/plugin-extensibility.md

```diff
@@ -0,0 +1,277 @@
+
+*payload* : Required field that contains payload to be signed. This is currently the JWSPayload that includes the subject descriptor and other JWS claims.
+
+*Response*
+All response attributes are required.
+
+```json
+{    
+  // The same key id as passed in the request.
+  "keyId" : <key id>,
+  "signature" : <Base64 encoded signature>,
+  "signingAlgorithm" : <signing algorithm>
+  "certificateChain": ["Base64(DER(leafCert))","Base64(DER(intermediateCACert))","Base64(DER(rootCert))"]
+}
+```
+
+*signingAlgorithm* : One of following [supported signing algorithms](https://github.com/notaryproject/notaryproject/blob/main/signature-specification.md#algorithm-selection), Notation uses this validate the signature, and to set the appropriate attribute in signature envelope (e.g. JWS `alg`). `RSASSA_PSS_SHA_256`, `RSASSA_PSS_SHA_384`, `RSASSA_PSS_SHA_512`,  `ECDSA_SHA_256`, `ECDSA_SHA_384`, `ECDSA_SHA_512`.
+
+*certificateChain* : Ordered list of certificates starting with leaf certificate and ending with root certificate.
+
+*Error codes*
+
+* VALIDATION_ERROR - Any of the required request fields was empty, or a value was malformed/invalid. Includes condition where the key referenced by `keyId` was not found.
+* UNSUPPORTED_CONTRACT_VERSION - The contract version used in the request is unsupported.
+* ACCESS_DENIED - Authentication/authorization error to use given key.
+* TIMEOUT - The operation to generate signature timed out and can be retried by Notation.
+* THROTTLED - The operation to generate signature was throttles and can be retried by Notation.
+* ERROR - Any general error that does not fall into previous error categories.
+
+### Signature Envelope Generator
+
+This interface targets plugins that in addition to signature generation want to generate the complete signature envelope. This interface allows plugins to have full control over the generated signature envelope, and can append additional signed and unsigned metadata, including timestamp signatures. The plugin must be signature envelope format aware, and implement new formats when Notary adopts new formats.
+
+#### Signing workflow using plugin
+
+1. Given a user request to sign `image`, with `key`
+1. Pull the image manifest using `image` url, and construct a descriptor
+1. Append any user provided metadata and Notary metadata as descriptor annotations.
+1. Determine if the registered key uses a plugin
+1. Execute the plugin with `discover` command
+    1. If plugin supports capability `SIGNATURE_ENVELOPE_GENERATOR`
+        1. Execute the plugin with `generate-envelope` command, set `request.key` to key definition (from `/notation/config.json`), `request.payload` to base64 encoded descriptor, `request.payloadType` to `application/vnd.oci.descriptor.v1+json` and `request.signatureEnvelopeType` to `application/vnd.cncf.notary.v2.jws.v1`.
```

Comment:
Will Notation generate timestamp signature for `Signature Envelope Generator` plugin or its responsibility of plugin publisher?

Reply:
Good point, will address in next rev.